fvm/environment/event_emitter.go

```diff
@@ -0,0 +1,207 @@
+package environment
+
+import (
+	"fmt"
+
+	"github.com/onflow/cadence"
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+
+	"github.com/onflow/flow-go/fvm/errors"
+	metering "github.com/onflow/flow-go/fvm/meter"
+	"github.com/onflow/flow-go/fvm/systemcontracts"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+)
+
+// EventEmitter collect events, separates out service events, and enforces
+// event size limits.
+//
+// Note that scripts do not emit events, but must expose the API in compliance
+// with the runtime environment interface.
+type EventEmitter interface {
+	// Cadence's runtime API.  Note that the script variant will return
+	// OperationNotSupportedError.
+	EmitEvent(event cadence.Event) error
+
+	Events() []flow.Event
+	ServiceEvents() []flow.Event
+}
+
+type NoEventEmitter struct{}
+
+func (NoEventEmitter) EmitEvent(event cadence.Event) error {
+	return errors.NewOperationNotSupportedError("EmitEvent")
+}
+
+func (NoEventEmitter) Events() []flow.Event {
+	return []flow.Event{}
+}
+
+func (NoEventEmitter) ServiceEvents() []flow.Event {
+	return []flow.Event{}
+}
+
+type eventEmitter struct {
+	tracer *Tracer
+	meter  Meter
+
+	chain   flow.Chain
+	txID    flow.Identifier
+	txIndex uint32
+	payer   flow.Address
+
+	serviceEventCollectionEnabled bool
+	eventCollectionByteSizeLimit  uint64
+	eventCollection               *EventCollection
+}
+
+// NewEventEmitter constructs a new eventEmitter
+func NewEventEmitter(
+	tracer *Tracer,
+	meter Meter,
+	chain flow.Chain,
+	txID flow.Identifier,
+	txIndex uint32,
+	payer flow.Address,
+	serviceEventCollectionEnabled bool,
+	eventCollectionByteSizeLimit uint64,
+) *eventEmitter {
+	return &eventEmitter{
+		tracer:                        tracer,
+		meter:                         meter,
+		chain:                         chain,
+		txID:                          txID,
+		txIndex:                       txIndex,
+		payer:                         payer,
+		serviceEventCollectionEnabled: serviceEventCollectionEnabled,
+		eventCollectionByteSizeLimit:  eventCollectionByteSizeLimit,
+		eventCollection:               NewEventCollection(),
+	}
+}
+
+func (emitter *eventEmitter) EventCollection() *EventCollection {
+	return emitter.eventCollection
+}
+
+func (emitter *eventEmitter) EmitEvent(event cadence.Event) error {
+	defer emitter.tracer.StartExtensiveTracingSpanFromRoot(
+		trace.FVMEnvEmitEvent).End()
+
+	err := emitter.meter.MeterComputation(metering.ComputationKindEmitEvent, 1)
+	if err != nil {
+		return fmt.Errorf("emit event failed: %w", err)
+	}
+
+	payload, err := jsoncdc.Encode(event)
+	if err != nil {
+		return errors.NewEncodingFailuref(
+			"failed to json encode a cadence event: %w",
+			err)
+	}
+
+	payloadSize := uint64(len(payload))
```

Comment:
q: do we consider using `json.Compact()` here? is `eventCollectionByteSizeLimit` for compacted or non-compacted json data, or is it really a don't care?

Reply:
my guess is they just needed some form of limiting.  the exact value doesn't matter